WordPress/gutenberg
# Automatic CSS Variable fallbacks (Components CSS)

<img width="403" alt="Screen Shot 2020-07-21 at 9 29 42 AM" src="https://user-images.githubusercontent.com/2322354/88062516-ee01f680-cb36-11ea-8a24-21f2989b577d.png">

(Image shows a generated CSS declaration fallback for it's `var()` usage in `RangeControl`)

This is a follow up to:
https://github.com/WordPress/gutenberg/pull/24059

This update adds automatic CSS variable fallback handling for CSS-in-JS generated styles from our `@wordpress/components` package (powered by [Emotion](https://emotion.sh/docs/introduction)).

This is accomplished by combining a couple of parts:

* A custom stylis plugin that can transform CSS declaration output
* A custom Emotion provider to target CSS-in-JS generated styles

### ✨ Stylis Plugin

[Stylis](https://github.com/thysultan/stylis.js) is the light-weight, power, and fast (and awesome) CSS compiler that many CSS-in-JS libraries use (e.g. Styled Components, Emotion, Linaria, etc...). Stylis allows us to add custom middleware (aka. plugins) that can modify output. This pathway is enabled by Emotion with the creation of a "custom cache".

### 👩‍🎤 Emotion Provider (and custom cache)

Don't let the custom cache scare you. This is how Emotion works under-the-hood! We're creating our own cache instance (the brains that coordinates everything for Emotion) and syncing it with the components via a Provider.

This is the same technique I introduced in this PR:
https://github.com/WordPress/gutenberg/pull/23215

(I'll need to update that one once this one merges)

### 🙏 IE Only Please

To optimize for performance, the transformations only occur on browsers that **do not support CSS variables**. The plugin auto-detects this by running a `window.CSS.supports()` check.

This option can be enabled/disabled by passing `skipSupportedBrowsers: false` when setting up the plugin:

```
const createCustomCache = memoize( () => {
	return createEmotionCache( {
		stylisPlugins: [
			stylisPluginCssCustomProperties( { skipSupportedBrowsers: false } ),
		],
	} );
} );
```

This is useful for **tests** (but also for debugging)

### 🍿 StyledProvider

This new `<StyledProvider />` enables us to seamlessly enhance our CSS-in-JS workflow by using additional plugins.

Some of these enhancements include...

* Automatic scope boosting (to ensure consistent + predictable CSS overrides)
* Automatic RTL handling

## How has this been tested?

Tested heavily with a barrage of unit tests. These unit tests throw wild (but valid) combinations of CSS variables, fallbacks, and odd value combinations to ensure that the parser works correctly.

The integration with the `<StyledProvider />` (and generated styles) were tested manually in local Gutenberg (for both the editor and Site Edit).

To test it locally...

* Add `skipSupportedBrowsers: false` to the `stylisPluginCssCustomProperties()` plugin
* Run `npm run dev`
* Add a Cover block with an image
* Inspect element on the `RangeSlider` component
* You should see a color fallback

<img width="403" alt="Screen Shot 2020-07-21 at 9 29 42 AM" src="https://user-images.githubusercontent.com/2322354/88064257-083cd400-cb39-11ea-88c2-c22eec89c03c.png">


## Types of changes

* Created a custom stylis plugin
* Created a Provider that hooks it up
* Integrates Provider with EditPost and EditSite
* Removed manual CSS variable fallbacks (that were recently added)

cc'ing @diegohaz <3 (Just in case you're interested)

## Checklist:
- [n/a] My code follows the accessibility standards.
